docs/hugo/content/design/ADR-2023-07-PackageReferences/_index.md

```diff
@@ -0,0 +1,175 @@
+
+<!-- yuml.me class diagram
+
+[<<interface>>;PackageReference]
+
+[PackageReference]<>--[LocalPackageReference]
+[PackageReference]<>--[StoragePackageReference]
+[PackageReference]<>--[ExternalPackageReference]
+
+[TypeName;Name string]--packageReference >[PackageReference]
+
+[TypeDefinition]--name >[TypeName]
+[TypeDefinition]--theType >[Type]
+
+[Type]<>--[TypeName]
+
+[<<interface>>;LocalLikePackageReference]<>--[LocalPackageReference]
+[LocalLikePackageReference]<>--[StoragePackageReference]
+
+[StoragePackageReference]-inner >[PackageReference]
+
+[PackageReference]<>--[SubPackageReference]
+[SubPackageReference]-parent >[PackageReference]
+
+-->
+
+* PRO: Somewhat tidies up the existing model.
+* CON: Still allows illegal states to be represented.
+* CON: Requires a type downcast in every case where we need to now use `LocalLikePackageReference` instead of `PackageReference`.
+
+### Option 3
+
+As a part of adding `SubPackageReference`, we might slightly rework the domain model to make illegal states impossible to represent.
+
+Not only will this be easier to understand, but it should streamline some parts of our existing codebase by removing the need to check for illegal states.
+
+{{< figure src="option-3.png" >}}
+
+<!-- yuml.me class diagram
+
+[<<interface>>;PackageReference]
+[PackageReference]<>--[LocalPackageReference]
+[PackageReference]<>--[StoragePackageReference]
+[PackageReference]<>--[SubPackageReference]
+
+[TypeName;Name string]--packageReference >[PackageReference]
+
+[TypeDefinition]--name >[TypeName]
+[TypeDefinition]--theType >[Type]
+
+[Type]<>--[TypeName]
+[Type]<>--[ExternalTypeName]
+
+[StoragePackageReference]-inner >[PackageReference]
+[SubPackageReference]-parent >[PackageReference]
+
+[ExternalTypeName;Name string]--packageReference >[ExternalPackageReference]
+
+-->
+
+We only ever need an `ExternalPackageReference` when naming a type that we import from elsewhere, so we make that a separate case, with a dedicated `ExternalTypeName` that can only reference such an external type.
+
+Breaking the implementation link between `ExternalPackageReference` and `PackageReference` means it's no longer possible for a `TypeDefinition` to be destined for an external package; nor is it possible for a StoragePackageReference to be based on the same. 
+
+The new implementation `SubPackageReference` declares a `parent` package for nesting.
+
+* PRO: Makes illegal states impossible to represent.
+* PRO: Removes need for type downcasts.
+* PRO: Removes need for checks for illegal states.
+* PRO: Makes the model easier to understand.
+* CON: Requires slightly more rework of existing code.
+
+### Option 4
+
+As for Option 4, but we also rename types to better reflect the new semantics. Given the introduction of `ExternalPackageReference`, maybe we use `InternalPackageReference`:
```

Comment:
What about TypeName, do we want to rename to `InternalTypeName`?

Reply:
I think I showed that on the diagram, so yes.